DOC-11959: Vector Search Code Snippets - iOS

velicuvlad opened the pull request.

velicuvlad committed:
Vector Search Code Snippets - iOS

velicuvlad committed:
typo

pasin commented on modules/objc/VectorSeachSnippets.m:
```diff
@@ -0,0 +1,120 @@
+    [collection createIndexWithName: @"vector_index" config: config error: &error];
+    
+    model = [NLEmbedding wordEmbeddingForLanguage:@"english"];
+    [model vectorForString: @"<word>"];
```
These two lines shouldn't be part of the code snippet.

pasin commented on modules/objc/VectorSeachSnippets.m:
```diff
@@ -0,0 +1,120 @@
+    CBLVectorIndexConfiguration* config = [[CBLVectorIndexConfiguration alloc] initWithExpression: @"vector" dimensions: 300 centroids: 20];
+    
+    // Default values already set:
+    NSLog(@"%@", config.encoding); // .scalarQuantizer(type: .SQ8)
```
I'm not sure if it will print out a useful info.

velicuvlad replied:
swift is fine.
obj-c is giving the type and the mem address... `CBLScalarQuantizer:<...>`. I couldn't find a more suitable way to show the default values - I am open for suggestions

pasin replied:
If we explain this in the doc content for this code snippet, we don't need to have all ofthe NSLog to show the default value at all.

velicuvlad replied:
I do agree

pasin replied:
I think it's better to describe it in the content instead of showing it as a comment in the code. So let remove it and add it back if we really need it.

pasin commented on modules/objc/VectorSeachSnippets.m:
```diff
@@ -0,0 +1,120 @@
+    config.encoding = [CBLVectorEncoding none];
+    config.metric = kCBLDistanceMetricCosine;
+    config.minTrainingSize = 50;
+    config.maxTrainingSize = 300;
```
If the scope of this code snippet is only for creating the default config, these 4 lines are not relevant.

velicuvlad replied:
those are to show how to set custom settings

ElliotFrancisHunter replied:
It is useful to show this, maybe specify with comments that these are custom settings, what they actually change for the end user etc

velicuvlad replied:
there's line 45 to mark this. I dont think it needs extra code comments within the snippet... plain english documentation should be enough... 

all this information will be available within the api doc as well anyway, so I am not sure if is needed to show "how to assign value to property" at all @pasin

pasin replied:
@ElliotFrancisHunter I'll let you decide on this as it needs to be inline with the content structure.

ElliotFrancisHunter replied:
@pasin @velicuvlad the current position is good for the purposes of getting the content usable on time for the release. Hakim and I intend to refactor how the mobile docs codebase handles snippets so it's a bit more sustainable in future so there's no need for movement currently

pasin commented on modules/objc/VectorSeachSnippets.m:
```diff
@@ -0,0 +1,120 @@
+    CBLQuery* query = [database createQuery: sql error: &error];
+    
+    CBLQueryParameters* parameters = [[CBLQueryParameters alloc] init];
+    [parameters setValue: <vectorArray> forName: @"vector"];
```
<vectorArray> will make the code snippet not being compiled. Also what is the scope of this code snippet? Does it need to show query as well?

velicuvlad replied:
from my understanding is showing how to get the vector from Embedding and then how to pass it as params to the query

velicuvlad replied:
I've address this by just using a declared placeholder

pasin commented on modules/objc/VectorSeachSnippets.m:
```diff
@@ -0,0 +1,120 @@
+    NSError* error;
+
+    model = [NLEmbedding wordEmbeddingForLanguage:@"english"];
+    [[CBLDatabase prediction] registerModel: model withName:@"WordEmbedding"];
```
This is incorrect. The model needs to implement CBLPredictiveModel protocol. Also, we will need to sync the code snippts for all platforms. I think we need some info or direction about the code snippet.

Anyway, here are two samples:

Samples:

// Color Model (I think server team has used RGB color as one of their sample apps)

```
@interface ColorModel : NSObject <CBLPredictiveModel>
@end

@implementation ColorModel

- (CBLDictionary*) predict: (CBLDictionary*)input {
   NSString* color = [input stringForKey: @"color"];
   NSArray* rgb = [RGBModel getRGBVectorForColorName: color]; // I made up the name of the model.
   
   CBLMutableDictionary* output = [[CBLMutableDictionary alloc] init];
   [output setValue: rgb forKey: @"vector"];
   return output;
}

@end
```

// Word Model 

```
@interface WordModel : NSObject <CBLPredictiveModel>
@end

@implementation WordModel

- (CBLDictionary*) predict: (CBLDictionary*)input {
   NSString* word = [input stringForKey: @"word"];
   NSArray* vector = [WordModel getVectorForWord: word]; // I made up the name of the model.
   
   CBLMutableDictionary* output = [[CBLMutableDictionary alloc] init];
   [output setValue: vector forKey: @"vector"];
   return output;
}

@end
```

velicuvlad replied:
@ElliotFrancisHunter do we want/need to align with server on this?

ElliotFrancisHunter replied:
Aligning with server on examples is a good idea if we have customers using multiple products. I'm in favour of this choice

velicuvlad replied:
and what is server doing then?

ElliotFrancisHunter replied:
Server seems to have some demo code here https://docs.couchbase.com/server/current/vector-search/create-vector-search-index-rest-api.html#example

which is related to colour. For mobile let's use the word examples as it's simpler for integration into MVP of beta docs. We can refine them over time with more detail such as the colour model

pasin replied:
PM suggested to me on the other day to have demo app converted to use the server's color model. I didn't have time to do it yet but the new sample app going forward will be color. It's ok to use word model as a sample now and convert it to use color later. It's not much of the change.

velicuvlad replied:
will use word model for now

velicuvlad committed:
re arrange

velicuvlad committed:
feedback

ElliotFrancisHunter reviewed (approved):
I'm happy with these snippets. Plenty of useful stuff for me to work with 👍🏼 As long as @pasin is happy I'll merge and add these to the docs.

pasin commented on modules/swift/VectorSearchSnippets.swift:
```diff
@@ -0,0 +1,113 @@
+        let config = VectorIndexConfiguration(expression: "word", dimensions: 300, centroids: 20)
+        try collection.createIndex(withName: "vector_index", config: config)
+        
+        model = NLEmbedding.wordEmbedding(for: .english)!
```
The code from this line is not relavent. There is also `useVectorMatch()` that shows the query.

pasin commented on modules/swift/VectorSearchSnippets.swift:
```diff
@@ -0,0 +1,113 @@
+    func vectorIndexPredictiveModel() throws {
+        // MARK: Create Vector Index with Predictive Model
+        let model = NLEmbedding.wordEmbedding(for: .english)!
+        Database.prediction.registerModel(model as! PredictiveModel, withName: "WordEmbedding")
```
Again this is wrong. Please fix this for Swift as well.

pasin commented on modules/objc/VectorSeachSnippets.m:
```diff
@@ -0,0 +1,143 @@
+    CBLVectorIndexConfiguration* config = [[CBLVectorIndexConfiguration alloc] initWithExpression: @"vector" dimensions: 300 centroids: 20];
+    [collection createIndexWithName: @"vector_index" config: config error: &error];
+    
+    model = [NLEmbedding wordEmbeddingForLanguage: @"english"];
```
The code from this line is not relavent. There is already `useVectorMatch()` sample.

pasin reviewed (commented):
I have just noticed this.

Should the code snippet is in the project here?

https://github.com/couchbase/docs-couchbase-lite/tree/release/3.1/modules/objc/examples
 
https://github.com/couchbase/docs-couchbase-lite/tree/release/3.1/modules/swift/examples
 
Also we have tried to make the code snippet compileable in those two XCode project. I think it's a good idea to follow that so that we don't make mistake in the code snippet.

pasin commented on modules/objc/VectorSeachSnippets.m:
```diff
@@ -0,0 +1,143 @@
+
+
+- (CBLDictionary*) predict: (CBLDictionary*)input {
+   
```
extra line

pasin commented on modules/objc/VectorSeachSnippets.m:
```diff
@@ -0,0 +1,143 @@
+}
+
+- (void) createVectorIndex {
+    NSError* error;
```
Move error to the line near `createIndexWithName`.

pasin commented on modules/objc/VectorSeachSnippets.m:
```diff
@@ -0,0 +1,143 @@
+- (void) createVectorIndex {
+    NSError* error;
+
+    WordModel* model;
```
model needs to be created.

```
WordModel* model = [[WordModel alloc] init];
```

pasin commented on modules/objc/VectorSeachSnippets.m:
```diff
@@ -0,0 +1,143 @@
+    
+    // MARK: Create a default Vector Index Configuration
+
+    CBLVectorIndexConfiguration* config = [[CBLVectorIndexConfiguration alloc] initWithExpression: @"vector" dimensions: 300 centroids: 20];
```
Change to use centorids = 8 for all for consistency.

pasin commented:
If this is a new location, can you create an XCode project and add the file into the project? So that we don't need to manually copy the code and compile it somewhere else when we make change to the code.

pasin commented:
You could aso add to the PR validation to verify the code snippet as well.

Here is an example in release/3.1 branch.
https://github.com/couchbase/docs-couchbase-lite/blob/release/3.1/.github/workflows/xcodebuild.yml

velicuvlad committed:
cleaned up and addressed feedback

move swift snippets to the xcode project and fix build issues

move obj-c snippets inside the xcode project and fix building issues

velicuvlad committed:
clean obj-c project

velicuvlad committed:
remove original placement

velicuvlad committed:
update actions

velicuvlad committed:
actions on PR instead of push

velicuvlad committed:
pick up the correct simulator in Actions

velicuvlad committed:
removed printing default values

velicuvlad committed:
removed workflow - will be moved on Jenkins

velicuvlad commented:
The validation automation must be done via Jenkins for multiple reasons such as access to the unreleased binaries etc. I have built the project on local and is fine.

This PR is ready for merge and can be merged if urgent until I find some time to take it with Jenkins

ElliotFrancisHunter commented:
> You could aso add to the PR validation to verify the code snippet as well.
> 
> Here is an example in release/3.1 branch. https://github.com/couchbase/docs-couchbase-lite/blob/release/3.1/.github/workflows/xcodebuild.yml
> 
> I'm asking this as in the past we made a lot of mistake in the code snippet.
> 
> You can have the structure like :
> 
> ```
> root/<codesnippet.swift>
> root/xcode/<Snippet.xcodeproject>
> ```
> 
> @ElliotFrancisHunter
> 
> Can we use the existing project to keep the code snippet? I doesn't need to use the tags. The goal is to have the code within the compliable project to make sure that the code snippet can be compiled without an error. We aso have the Github PR validation to check that the code snippet is compiled. We do have the problems in the past that the code snippet had a lot of mistake and was not compilable.

Yes that's fine. It'll make it easier for me to implement the snippets properly rather than the rough copy paste job I've done for sake of time.

ElliotFrancisHunter commented:
@pasin good to merge for sake of release 👍🏼 just waiting on your review

pasin commented on modules/swift/examples/code_snippets/VectorSearch.swift:
```diff
@@ -0,0 +1,130 @@
+    // MARK: Create Vector Index with Predictive Model
+
+    class WordModel: PredictiveModel {
+        
```
extra line

pasin commented on modules/swift/examples/code_snippets/VectorSearch.swift:
```diff
@@ -0,0 +1,130 @@
+        config.metric = .cosine
+        config.minTrainingSize = 50
+        config.maxTrainingSize = 300
+        
```
extra line

pasin commented on modules/objc/examples/code_snippets/VectorSeach.m:
```diff
@@ -0,0 +1,142 @@
+ */
+
+- (void) createDefaultIndex {
+    
```
extra line

pasin commented on modules/objc/examples/code_snippets/VectorSeach.m:
```diff
@@ -0,0 +1,142 @@
+}
+
+- (CBLQueryResultSet*) vectorIndexEmbedding {
+    
```
extra line

pasin commented on modules/objc/examples/code_snippets/VectorSeach.m:
```diff
@@ -0,0 +1,142 @@
+}
+
+- (CBLQueryResultSet*) useVectorDistance {
+    
```
extra line

pasin commented on modules/objc/examples/code_snippets/VectorSeach.m:
```diff
@@ -0,0 +1,142 @@
+}
+
+- (CBLQueryResultSet*) useVectorMatch {
+    
```
extra line

pasin commented on modules/swift/examples/code_snippets/VectorSearch.swift:
```diff
@@ -0,0 +1,130 @@
+
+
+
+
```
extra lines

pasin reviewed (approved):
@velicuvlad I have some minor comments about the extra lines. The rest looks good to me.

velicuvlad committed:
remove empty lines

velicuvlad commented:
@ElliotFrancisHunter ready
